### PR TITLE
Add option to blacklist URLs

### PR DESCRIPTION
**Pull request by ghost1227**
Fixes issue #851

### CODE REVIEW

File: includes/class-tracking.php [hunk 1]
```diff
@@ -368,19 +370,21 @@ public function fallback_track_visit() {
 		$affiliate_id = absint( $affiliate_id );
 
 		if( $this->is_valid_affiliate( $affiliate_id ) && ! $this->get_visit_id() ) {
+			if( ( ! empty( $_SERVER['HTTP_REFERER'] ) && ! affwp_is_url_banned( sanitize_text_field( $_POST['referrer'] ) ) ) || empty( $_SERVER['HTTP_REFERER'] ) ) {
```
Comment (pippinsplugins) [inline includes/class-tracking.php]: Why are you checking the referrer in two different ways here?

File: includes/referral-functions.php [hunk 2]
```diff
@@ -232,3 +232,46 @@ function affwp_count_referrals( $affiliate_id = 0, $status = array(), $date = ar
...
+
+/**
+ * Determines if a URL is banned
+ *
+ * @since 1.7.5
+ * @return bool True if banned, false otherwise
+ */
+function affwp_is_url_banned( $url = '' ) {
+	if( empty( $url ) ) {
+		return false;
```
Comment (pippinsplugins) [inline includes/referral-functions.php]: all return statements should go through the final filter at the end of the function.

File: includes/class-tracking.php (hunk 1 above)
Comment (ghost1227) [inline includes/class-tracking.php]: I'm not... technically. The next code block allows requests with NO referrer to be tracked. If I didn't add the `|| empty( $_SERVER['HTTP_REFERER'] )` line, all requests that don't pass that referrer will get automatically blocked.

File: includes/referral-functions.php (hunk 2 above)
Comment (ghost1227) [inline includes/referral-functions.php]: Ok. I just copied the EDD email_banned function... We might want to update that too ;)

Comment (ghost1227) [inline includes/referral-functions.php]: Fixed

File: includes/class-tracking.php (hunk 1 above)
Comment (pippinsplugins) [inline includes/class-tracking.php]: Ok will test more

**Comment by DrewAPicture** [comment]
Re-branched from `release/1.9` and closing in favor of #1379.

**Pull request closed without merging**